ports/esp32/esp32_sdcard.c

```diff
@@ -0,0 +1,386 @@
+} sdcard_card_obj_t;
+
+#define SDCARD_CARD_FLAGS_HOST_INIT_DONE 0x01
+#define SDCARD_CARD_FLAGS_CARD_INIT_DONE 0x02
+
+#define _SECTOR_SIZE(self) (self->card.csd.sector_size)
+
+STATIC esp_err_t check_esp_err(esp_err_t code) {
+    const mp_obj_type_t *ex_type = NULL;
+
+    switch(code) {
+    case ESP_OK:
+        return ESP_OK;
+    case ESP_ERR_NO_MEM:
+        ex_type = &mp_type_MemoryError;
+        break;
+    case ESP_ERR_TIMEOUT:
+        ex_type = &mp_type_TimeoutError;
+        break;
+    case ESP_ERR_NOT_SUPPORTED:
+        ex_type = &mp_type_NotImplementedError;
+        break;
+    default:
+        mp_raise_OSError(code);
+    }
+
+    mp_raise_msg(ex_type, "From ESP SDK");
+}
+
+STATIC gpio_num_t pin_or_int(const mp_obj_t arg) {
+    if (mp_obj_is_small_int(arg))
+        return MP_OBJ_SMALL_INT_VALUE(arg);
+    else
+        // This raises a value error if the argument is not a Pin.
+        return machine_pin_get_id(arg);
+}
+
+#define SET_CONFIG_PIN(config, pin_var, arg_id) \
+    if (arg_vals[arg_id].u_obj != mp_const_none) \
+        config.pin_var = pin_or_int(arg_vals[arg_id].u_obj)
+
+STATIC esp_err_t sdcard_ensure_card_init(sdcard_card_obj_t *self, bool force) {
+    if (force || !(self->flags & SDCARD_CARD_FLAGS_CARD_INIT_DONE)) {
+        DEBUG_printf("  Calling card init");
+
+        esp_err_t err = sdmmc_card_init(&(self->host), &(self->card));
+        if (err == ESP_OK) {
+            self->flags |= SDCARD_CARD_FLAGS_CARD_INIT_DONE;
+        } else {
+            self->flags &= ~SDCARD_CARD_FLAGS_CARD_INIT_DONE;
+        }
+        DEBUG_printf("  Card init returned: %i", err);
+
+        return err;
+    } else {
+        return ESP_OK;
+    }
+}
+
+/******************************************************************************/
+// MicroPython bindings
+//
+// Expose the SD card or MMC as an object with the block protocol.
+
+// Create a new SDCard object
+// The driver supports either the host SD/MMC controller (default) or SPI mode
+// In both cases there are two "slots". Slot 0 on the SD/MMC controller is
+// typically tied up with the flash interface in most ESP32 modules but in
+// theory supports 1, 4 or 8-bit transfers. Slot 1 supports only 1 and 4-bit
+// transfers. Only 1-bit is supported on the SPI interfaces.
+// card = SDCard(slot=1, width=None, present_pin=None, wp_pin=None)
+
+STATIC mp_obj_t esp32_sdcard_make_new(const mp_obj_type_t *type, size_t n_args, size_t n_kw, const mp_obj_t *args) {
+    // check arguments
+    enum {
+        ARG_slot,
+        ARG_spi,
+        ARG_width,
+        ARG_cd_pin,
+        ARG_wp_pin,
+        ARG_miso_pin,
+        ARG_mosi_pin,
+        ARG_sck_pin,
+        ARG_cs_pin,
+    };
+    STATIC const mp_arg_t allowed_args[] = {
+        { MP_QSTR_slot,     MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = 1} },
+        { MP_QSTR_spi,      MP_ARG_KW_ONLY | MP_ARG_BOOL,{.u_bool = false} },
+        { MP_QSTR_width,    MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = 1} },
```

Comment:
Yet another alternative to having `spi`: this `width` argument could be None (or something else sensible, maybe even a constant like `SPI_WIDTH`) to select SPI.

Reply:
I think that it would be better to consider the two slots using that SD/MMC hardware and the two SPI interfaces to represent four slots, as suggested above. For the SPI slots the width will be ignored.